code/modules/jobs/job_exp.dm

```diff
@@ -0,0 +1,333 @@
+GLOBAL_LIST_EMPTY(exp_to_update)
+GLOBAL_PROTECT(exp_to_update)
+
+// Admin Verbs
+/client/proc/cmd_admin_check_player_exp()	//Allows admins to determine who the newer players are.
+	set category = "Admin"
+	set name = "Check Player Playtime"
+	if(!check_rights(R_ADMIN))
+		return
+	var/list/msg = list()
+	msg += "<html><head><title>Playtime Report</title></head><body>Playtime:<BR><UL>"
+	for(var/client/C in GLOB.clients)
+		msg += "<LI> - [key_name_admin(C)]: <A href='?_src_=holder;getplaytimewindow=\ref[C.mob]'>" + C.get_exp_living() + "</a></LI>"
+	msg += "</UL></BODY></HTML>"
+	src << browse(msg.Join(), "window=Player_playtime_check")
+
+/datum/admins/proc/cmd_show_exp_panel(client/C)
+	if(!C)
+		to_chat(usr, "ERROR: Client not found.")
+		return
+	if(!check_rights(R_ADMIN))
+		return
+	var/list/body = list()
+	body += "<html><head><title>Playtime for [C.key]</title></head><BODY><BR>Playtime:"
+	body += C.get_exp_report()
+	body += "<A href='?_src_=holder;toggleexempt=\ref[C]'>Toggle Exempt status</a>"
+	body += "</BODY></HTML>"
+	usr << browse(body.Join(), "window=playerplaytime[C.ckey];size=550x615")
+
+
+/datum/admins/proc/toggle_exempt_status(client/C)
+	if(!C)
+		to_chat(usr, "ERROR: Client not found.")
+		return
+	if(!check_rights(R_ADMIN))
+		return
+
+	set_db_player_flags(C)
+	var/dbflags = C.prefs.db_flags
+	var/newstate = FALSE
+	if(dbflags & DB_FLAG_EXEMPT)
+		newstate = FALSE
+	else
+		newstate = TRUE
+
+	message_admins("[key_name_admin(usr)] has [newstate ? "activated" : "deactivated"] job exp exempt status on [key_name_admin(C)]")
+	update_flag_db(C,DB_FLAG_EXEMPT, newstate)
+
+// Procs
+
+
+/datum/job/proc/required_playtime_remaining(client/C)
+	if(!C)
+		return 0
+	if(!config.use_exp_tracking)
+		return 0
+	if(!exp_requirements || !exp_type)
+		return 0
+	if(!job_is_xp_locked(src.title))
+		return 0
+	if(config.use_exp_restrictions_admin_bypass && check_rights(R_ADMIN, 0, C.mob))
+		return 0
+	var/isexempt = C.prefs.db_flags & DB_FLAG_EXEMPT
+	if(isexempt)
+		return 0
+	var/my_exp = calc_exp_type(C,get_exp_req_type())
+	var/job_requirement = get_exp_req_amount()
+	if(my_exp >= job_requirement)
+		return 0
+	else
+		return (job_requirement - my_exp)
+
+/datum/job/proc/get_exp_req_amount()
+	if(title in GLOB.command_positions)
+		if(config.use_exp_restrictions_heads_hours)
+			return config.use_exp_restrictions_heads_hours * 60
+	return exp_requirements
+
+/datum/job/proc/get_exp_req_type()
+	if(title in GLOB.command_positions)
+		if(config.use_exp_restrictions_heads_department && exp_type_department)
+			return exp_type_department
+	return exp_type
+
+/proc/job_is_xp_locked(jobtitle)
+	if(!config.use_exp_restrictions_heads && jobtitle in GLOB.command_positions)
+		return FALSE
+	if(!config.use_exp_restrictions_other && !(jobtitle in GLOB.command_positions))
+		return FALSE
+	return TRUE
+
+/proc/calc_exp_type(client/C,exptype)
+	var/list/explist = C.prefs.exp.Copy()
```

Comment:
Can I just say I REALLY dislike exp being stored in the preference datum. Why not just store it on the client?